src/main/java/com/dabsquared/gitlabjenkins/connection/GitLabConnection.java

```diff
@@ -25,15 +40,30 @@
     private transient String apiToken;
     // TODO make final when migration code gets removed
     private String apiTokenId;
+    private GitLabClientBuilder clientBuilder;
     private final boolean ignoreCertificateErrors;
     private final Integer connectionTimeout;
     private final Integer readTimeout;
+    private transient GitLabClient apiCache;
 
     @DataBoundConstructor
-    public GitLabConnection(String name, String url, String apiTokenId, boolean ignoreCertificateErrors, Integer connectionTimeout, Integer readTimeout) {
+    public GitLabConnection(String name, String url, String apiTokenId, String clientBuilderId, boolean ignoreCertificateErrors, Integer connectionTimeout, Integer readTimeout) {
```

Comment:
👎 as this breaks backward compatibility. The old constructor should still exist, and call the new constructor.

Reply:
You're right, this will break backward-compatiblity. But only in those case where someone outside of this plugin creates a GitLabConnection programmatically instead of using the configuration page and then getting it in a build via GitLabConnectionProperty#getClient().

Is this class is intentionally part of the public api of this plugin (and not just because jenkins needs it to be to do it's magic)?  I'd say the only ways for anyone "outside" of this plugin to get a connection/api-client should be via GitLabConnectionProperty#getClient() or by looking up the configured connections by enumerating the GitLabConnectionConfigs and then calling GitLabConnectionConfig#getClient(). 

So instead of making an effort to keep compatibility here, i'd rather annotate at least the constructors with @Restricted(NoExternalUse.class) so that they're only accessible outside of the plugin.

Or is there something i'm missing?

Oh, i've already added code to GitLabConnection#readResolve that ensures existing connections are upated to use the autodetecting client.

Reply:
> You're right, this will break backward-compatiblity. But only in those case where someone outside of this plugin creates a GitLabConnection programmatically.

Right and that's the problem. There's no guarantee that someone has not created a groovy script which creates a Jenkins jobs and attaches a `GitLabConnection` using the old constructor. If you were to remove this constructor then that would mean that the person must update their scripts which might take some time, or they would have to do more work to figure out what happened. 

Here is an example where someone uses the constructor you're going to remove: 

https://github.com/rabits/jenkinsbro/blob/d5d279e340de638d6e5586292fd04ec5456e66a0/jenkinsBro/lib/modules/gitlab.groovy#L9-L14

This is just one example but someone might write a script for their non-public groovy scripts which will break, and that would mean that they need to update their scripts.

I would like to merge the PR which uses `@DataBoundSetter` annotations as this creates one empty constructor and everything is added using setter methods.

> Is this class intentionally part of the public api of this plugin (and not just because jenkins needs it to be to do it's magic)?

Once a method becomes public it's usually difficult to change this. It has nothing to do with any magic, but it's how public apis are. Once they're public it should be deprecated if they shouldn't be used.

Reply:
So i'll keep the old constructor, deprecate it and move the @DateBoundConstructor-annotion to the new constructor (as there may only be one) and add a restricting annotion the prevent it leaking into the public api - except these constructors should be part of the public api.

Btw: with "jenkins-magic" i mean, that the class must be public because otherwise jenkins is not able to detect/call the @DataBoundConstructor.